pkg/visitors/table.go

```diff
@@ -26,57 +31,105 @@ func (v *Table) Run(f uefi.Firmware) error {
+		return v.printRow(f, fmt.Sprintf("%T", f), "", "", 0, 0)
 	}
 }
 
 func indent(n int) string {
 	return strings.Repeat(" ", n)
 }
 
-func (v *Table) printRow(f uefi.Firmware, node, name, typez interface{}) error {
+func (v *Table) printRow(f uefi.Firmware, node, name, typez interface{}, offset, dataOffset uint64) error {
 	if v.W == nil {
 		v.W = tabwriter.NewWriter(os.Stdout, 0, 0, 2, ' ', 0)
 		defer func() { v.W.Flush() }()
-		fmt.Fprintf(v.W, "%sNode\tGUID/Name\tType\tSize\n", indent(v.indent))
+		if v.Layout {
+			fmt.Fprintf(v.W, "%sNode\tGUID/Name\tOffset\tSize\n", indent(v.indent))
+		} else {
+			fmt.Fprintf(v.W, "%sNode\tGUID/Name\tType\tSize\n", indent(v.indent))
+		}
+	}
+	length := uint64(len(f.Buf()))
+	if typez == "" {
+		if uefi.IsErased(f.Buf(), uefi.Attributes.ErasePolarity) {
+			typez = "(empty)"
+		}
+	}
+	if v.Layout {
+		if name == "" {
+			name = typez
+		}
+		fmt.Fprintf(v.W, "%s%v\t%v\t%#08x\t%#08x\n", indent(v.indent), node, name, offset, length)
+	} else {
+		fmt.Fprintf(v.W, "%s%v\t%v\t%v\t%#8x\n", indent(v.indent), node, name, typez, length)
 	}
-	fmt.Fprintf(v.W, "%s%v\t%v\t%v\t%#8x\n", indent(v.indent), node, name, typez, len(f.Buf()))
 	v2 := *v
 	v2.indent++
-	if err := f.ApplyChildren(&v2); err != nil {
-		return err
+	v2.offset = dataOffset
+	v2.curOffset = v2.offset
+	if !v.TopLevel || v.indent < v.depth {
+
+		if err := f.ApplyChildren(&v2); err != nil {
+			return err
+		}
 	}
+	v.curOffset += length
 	switch f := f.(type) {
 	case *uefi.FirmwareVolume:
 		// Print free space at the end of the volume
-		fmt.Fprintf(v.W, "%s%v\t%v\t%v\t%#8x\n", indent(v2.indent), "Free", "", "", f.FreeSpace)
+		if v.Layout {
+			fmt.Fprintf(v.W, "%s%v\t%v\t%#08x\t%#08x\n", indent(v2.indent), "Free", "", offset+length-f.FreeSpace, f.FreeSpace)
+		} else {
+			fmt.Fprintf(v.W, "%s%v\t%v\t%v\t%#8x\n", indent(v2.indent), "Free", "", "", f.FreeSpace)
+		}
 	case *uefi.NVarStore:
 		// Print free space and GUID store
-		fmt.Fprintf(v.W, "%s%v\t%v\t%v\t%#8x\n", indent(v2.indent), "Free", "", "", f.GUIDStoreOffset-f.FreeSpaceOffset)
-		fmt.Fprintf(v.W, "%s%v\t%v\t%v\t%#8x\n", indent(v2.indent), "GUIDStore", "", fmt.Sprintf("%d GUID", len(f.GUIDStore)), f.Length-f.GUIDStoreOffset)
+		if v.Layout {
+			fmt.Fprintf(v.W, "%s%v\t%v\t%#08x\t%#08x\n", indent(v2.indent), "Free", "", offset+f.FreeSpaceOffset, f.GUIDStoreOffset-f.FreeSpaceOffset)
+			fmt.Fprintf(v.W, "%s%v\t%v\t%#08x\t%#08x\n", indent(v2.indent), "GUIDStore", fmt.Sprintf("%d GUID", len(f.GUIDStore)), offset+f.GUIDStoreOffset, f.Length-f.GUIDStoreOffset)
+		} else {
+			fmt.Fprintf(v.W, "%s%v\t%v\t%v\t%#8x\n", indent(v2.indent), "Free", "", "", f.GUIDStoreOffset-f.FreeSpaceOffset)
+			fmt.Fprintf(v.W, "%s%v\t%v\t%v\t%#8x\n", indent(v2.indent), "GUIDStore", "", fmt.Sprintf("%d GUID", len(f.GUIDStore)), f.Length-f.GUIDStoreOffset)
+		}
+	case *uefi.File:
+		// Align
+		// TODO: do we need the complex align logic from assemble?
```

Comment:
You don't need it as long as the user has assembled the image prior, or has not made any modifications to the tree. I haven't forced an assemble unconditionally because that makes it slow though, but there runs the risk that if a user doesn't assemble after removing, table may display old/incorrect information.

Reply:
Ok, I'll drop the TODO then thanks